### PR TITLE
feat: search improvements for #404 and #397 and #319

### PR DESCRIPTION
**Pull request by tomaszandel**
- search improvement
- tested on http://149.156.182.2:8983/solr and with NON facade recommender.

-goal: set KeywordTokenizer and add simple algo for KeywordTokenizer with ~2 maximum if more than 3 words with words size greater than 5. otherwise no fuzzy. case for ~1 also to be crafted. Important with KeywordTokenizer would be to add mm at 70-80%.

### CODE REVIEW

**Comment by agpul** [comment]
Use case: Datasource: "MetalPDB: a database of metal-binding sites in 3D structures of biological macromolecules"
1. Type: "a database of metal-binding"
- metal-binding is changed to metalbinding and it is not found
- the word "a" causes a lot of problems because the search engine is looking for single "a"
image

**Comment by agpul** [comment]
> Use case: Datasource: "MetalPDB: a database of metal-binding sites in 3D structures of biological macromolecules"
> 
> 1. Type: "a database of metal-binding"
> 
> * metal-binding is changed to metalbinding and it is not found
> * the word "a" causes a lot of problems because the search engine is looking for single "a"
>   image

maybe we should not take into consideration short words like "a", "of" etc... is it possible?

**Comment by agpul** [comment]
Remove: 
image

**Comment by agpul** [comment]
Just a question: 
image
What do you think, what should be more important: August x2 or "August 1964" - for me "August 1964":)

**Comment by agpul** [comment]
> Just a question: image What do you think, what should be more important in this case: August x2 or "August 1964" ?- for me "August 1964":)

Or maybe we should add the support for ""? In case of such typing: 
image
we should have "August 1964" on the top.

**Comment by eldenkat** [comment]
###

**Comment by agpul** [comment]
> 

@eldenkat what is the problem here?

**Comment by agpul** [comment]
I type: Sleeping site selection in two Asian viverrids: effects of predation risk, resource access and habitat characteristics"

why "habitat characteristics" is not highlighted in the title but is highlighted in the description?:)
image

**Comment by agpul** [comment]
> > 
> 
> @eldenkat what is the problem here? Seems odd, no name, no title, three times the same position. What does it say to the user?

I think that it is not a scope of this issue because the search engine works properly - it is not the blame of the search engine it is the blame of OpenAire data, but @tomaszandel can confirm or reject:P

**Comment by eldenkat** [comment]
Do author and title carry the same importance when searching by typed word? @agpul

**Comment by eldenkat** [comment]
make the clear all button opaque?

**Review by eldenkat — CHANGES_REQUESTED** [review]
Pagination does not work when results are listed on more than one page

**Comment by eldenkat** [comment]
No highlights in trainings. Other categories shows higlights in the same search case.

**Comment by eldenkat** [comment]
SS does not display a search result for the query from the first item in the list.

**Comment by Marcelinna** [comment]
Why [Bouchaib Bahli] is higer than [Maksi Bouchaib]
image

**Comment by Marcelinna** [comment]
I am not sure that autocomplete works fine
image

**Comment by tomaszandel** [comment]
> Just a question: image What do you think, what should be more important in this case: August x2 or "August 1964" ?- for me "August 1964":)

Yes, this is why keywordTokenizer is probably better idea, it shoudl match whole sentences better,  but in some other cases it will lack search flexibility ;)

**Comment by tomaszandel** [comment]
> > Use case: Datasource: "MetalPDB: a database of metal-binding sites in 3D structures of biological macromolecules"
> > 
> > 1. Type: "a database of metal-binding"
> > 
> > 
> > * metal-binding is changed to metalbinding and it is not found
> > * the word "a" causes a lot of problems because the search engine is looking for single "a"
> >   image
> 
> maybe we should not take into consideration short words like "a", "of" etc... is it possible?

Solr is returning that values properly, bug is on our side, we match the regex wrongly, thus it highlights every occurence, fix TODO

**Comment by tomaszandel** [comment]
> > Just a question: image What do you think, what should be more important in this case: August x2 or "August 1964" ?- for me "August 1964":)
> 
> Or maybe we should add the support for ""? In case of such typing: image we should have "August 1964" on the top.

hmm in fact it is correct according to our setup, 2 matched names gets higher score, so its technically OK. TO fix it we should consider a separate call like you mention " " that looks for 100% whole sentence match. Or just field selection where search would look for

**Comment by tomaszandel** [comment]
> > Use case: Datasource: "MetalPDB: a database of metal-binding sites in 3D structures of biological macromolecules"
> > 
> > 1. Type: "a database of metal-binding"
> > 
> > 
> > * metal-binding is changed to metalbinding and it is not found
> > * the word "a" causes a lot of problems because the search engine is looking for single "a"
> >   image
> 
> maybe we should not take into consideration short words like "a", "of" etc... is it possible?

- we can fix, we should make a list of separators we want to let in, when i remove - search works properly

**Comment by tomaszandel** [comment]
> I type: Sleeping site selection in two Asian viverrids: effects of predation risk, resource access and habitat characteristics"
> 
> why "habitat characteristics" is not highlighted in the title but is highlighted in the description?:) image

This is interesting case, because solr indeed is not highlighting it in title, but its a bit more complicated how he chooses what and what not to highlight, we could fix it on our end as we are using regex match, however why solr is not highlighting it is a mystery :)

**Comment by tomaszandel** [comment]
> Do author and title carry the same importance when searching by typed word? @agpul image image
> 
> which result should be higher? or should have the same weight

Yeah this is tricky, if want to locate by author names, author has to be boosted higher, otherwise simple names in many cases exists in titles, same with surnames. So its a tradeoff, ideally we should have a field selection in search that would fix everything

**Comment by tomaszandel** [comment]
> No highlights in trainings. Other categories show higlights in the same search case.
> 
> 

yeah it seems its a case not handled on our end, trainings are not highlighed at all, its a valid bug, solr replies correctly here

**Comment by tomaszandel** [comment]
> I am not sure that autocomplete works fine image

It is ok, remember that this is not really an autocomplete, because it searches title, descrition, authors, keywords and so on, if we want to have a real autocomplete for titles than we should search only in titles, or display 'autocomplete' tool differently, so we can see what is matching, here we show only titles, but matches could be in some other place, 

Zrzut ekranu 2023-02-7 o 23 14 29

**Comment by Marcelinna** [comment]
I type "INDIGO Identity and Access Management (IAM)", when I am on page 1 
image
I type "INDIGO Identity and Access Management (IAM)", when I am on page 6
image

File: backend/app/routes/web/search.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ async def search_post(
                 cursor=cursor,
                 facets=request.facets,
             )
-        except TransportError as e:
+        except Exception as e:
```
Comment (Michal-Kolomanski) [inline backend/app/routes/web/search.py:56]: Too broad exception handling. If `TransportError` is not enough, let's add others

File: transform/Pipfile [hunk 2]
```diff
@@ -18,4 +18,3 @@ black = {extras = ["jupyter"], version = "*"}
 
 [requires]
 python_version = "3.10"
-python_full_version = "3.10.5"
```
Comment (Michal-Kolomanski) [inline transform/Pipfile:21]: Please revert those changes, they are not in the scope

File: transform/Pipfile.lock [hunk 3]
```diff
@@ -1,11 +1,10 @@
 {
     "_meta": {
         "hash": {
-            "sha256": "b05676247bc1bee62f02110fb165a64718b360e4ad16a64c39a16bb1f9221d46"
+            "sha256": "7fa9536461fcb4880f7c56ec5fc6bb3dea620dd7cfda91fd123d8b1a4b79989a"
```
Comment (Michal-Kolomanski) [inline transform/Pipfile.lock:4]: Please revert those changes, they are not in the scope

**Review by Michal-Kolomanski — APPROVED** [review]
LGTM

**Comment by eldenkat** [comment]
Why a strict match doesn't go first?

**Comment by eldenkat** [comment]
Highlights.
Entering "horizontal service", page 1-4 highlights word "service", page 6 - no higlights and than returning to the page one by one 5,4,3,2,1 - no highlights.

**Review by Marcelinna — APPROVED** [review]

**Review by eldenkat — APPROVED** [review]

**Pull request merged**